sat_vapor_pres/sat_vapor_pres_k.F90

```diff
@@ -1148,580 +2053,1478 @@ end subroutine compute_mrs_k_0d
+   end if
 
    nbad = 0
-   do j = 1, size(temp,2)
-   do i = 1, size(temp,1)
-     tmp = temp(i,j)-tminl
-     ind = int(dtinvl*(tmp+tepsl))
-     if (ind < 0 .or. ind >= table_siz)  then
-       nbad = nbad+1
-     else
-       del = tmp-dtres*real(ind)
-       esat(i,j) = TABLE2(ind+1) + &
-                   del*(DTABLE2(ind+1) + del*D2TABLE2(ind+1))
-       desat(i,j) = DTABLE2(ind+1) + 2.*del*D2TABLE2(ind+1)
-     endif
-   enddo
-   enddo
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+     select type (esat)
+     type is (real(kind=r4_kind))
+       select type (desat)
+       type is (real(kind=r4_kind))
+         do j = 1, size(temp,2)
+         do i = 1, size(temp,1)
+           tmp = temp(i,j)-tminl
+           ind = int(dtinvl*(tmp+tepsl))
+           if (ind < 0 .or. ind >= table_siz)  then
+             nbad = nbad+1
+           else
+             del = tmp-dtres*real(ind)
+             esat(i,j) = TABLE2(ind+1) + del*(DTABLE2(ind+1)+del*D2TABLE2(ind+1))
+             desat(i,j) = DTABLE2(ind+1) + 2.*del*D2TABLE2(ind+1)
+           endif
+         enddo
+         enddo
+       end select
+     end select
+   type is (real(kind=r8_kind))
+     select type (esat)
+     type is (real(kind=r8_kind))
+       select type (desat)
+       type is (real(kind=r8_kind))
+         do j = 1, size(temp,2)
+         do i = 1, size(temp,1)
+           tmp = temp(i,j)-tminl
+           ind = int(dtinvl*(tmp+tepsl))
+           if (ind < 0 .or. ind >= table_siz)  then
+             nbad = nbad+1
+           else
+             del = tmp-dtres*real(ind)
+             esat(i,j) = TABLE2(ind+1) + del*(DTABLE2(ind+1)+del*D2TABLE2(ind+1))
+             desat(i,j) = DTABLE2(ind+1) + 2.*del*D2TABLE2(ind+1)
+           endif
+         enddo
+         enddo
+       end select
+     end select
+   end select
 
  end subroutine lookup_es2_des2_k_2d
 
 !#######################################################################
 
  subroutine lookup_es2_des2_k_1d (temp, esat, desat, nbad)
- real, intent(in),  dimension(:)  :: temp
- real, intent(out), dimension(:)  :: esat, desat
+ class(*), intent(in),  dimension(:)  :: temp
+ class(*), intent(out), dimension(:)  :: esat, desat
  integer, intent(out)             :: nbad
 
  real    :: tmp, del
  integer :: ind, i
+ logical :: valid_types = .false.
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+     select type (esat)
+     type is (real(kind=r4_kind))
+       select type (desat)
+       type is (real(kind=r4_kind))
+         valid_types = .true.
+       end select
+     end select
+   type is (real(kind=r8_kind))
+     select type (esat)
+     type is (real(kind=r8_kind))
+       select type (desat)
+       type is (real(kind=r8_kind))
+         valid_types = .true.
+       end select
+     end select
+   end select
+   if ( .not. valid_types ) then
+     call error_mesg ('sat_vapor_pres_k_mod::lookup_es2_des2_k_1d', 'unsupported kind', FATAL)
```

Comment:
This error message should be more descriptive.  *"temp, desat, and esat types do not match .or. desat and/or esat are not one of the supported types of real(kind=4) or real(kind=8)."*